clientcontroller/evm_consumer.go

```diff
@@ -81,27 +86,167 @@ func (ec *EVMConsumerController) QueryFinalityProviderVotingPower(fpPk *btcec.Pu
+		}
+		blocks = append(blocks, ib)
+	}
+	return blocks, nil
+
 }
 
 func (ec *EVMConsumerController) QueryBlocks(startHeight, endHeight, limit uint64) ([]*types.BlockInfo, error) {
 
-	return ec.queryLatestBlocks(sdk.Uint64ToBigEndian(startHeight), 0, finalitytypes.QueriedBlockStatus_ANY, false)
+	if endHeight < startHeight {
+		return nil, fmt.Errorf("the startHeight %v should not be higher than the endHeight %v", startHeight, endHeight)
+	}
+	count := endHeight - startHeight
+	if count > limit {
+		count = limit
+	}
+
+	lastnumber, err := ec.GetLatestFinalizedNumber()
+	if err != nil {
+		return nil, fmt.Errorf("can't get latest finalized block:%s", err)
+
+	}
+	type Block struct {
+		Number string
+		Hash   string
+	}
+
+	startnumber := new(big.Int).SetUint64(startHeight)
+
+	var B []rpc.BatchElem
+
+	for i := 0; i < int(count); i++ {
+
+		hexStr := Transform(startnumber)
+		ib := rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{hexStr, true},
+			Result: new(Block),
+		}
+		B = append(B, ib)
+		startnumber.Add(startnumber, big.NewInt(1))
+	}
+
+	err = ec.evmClient.BatchCall(B)
+	if err != nil {
+		return nil, fmt.Errorf("can't get  blocks")
+
+	}
+	var blocks []*types.BlockInfo
+
+	for _, b := range B {
+		nb := b.Result.(*Block)
+		num, err := strconv.ParseUint(strings.TrimPrefix(nb.Number, "0x"), 16, 64)
+		if err != nil {
+			return nil, fmt.Errorf("error:%s", err)
+		}
+
+		number := new(big.Int).SetUint64(num)
+		var finalized bool = false
+		if number.Cmp(lastnumber) <= 0 {
+			finalized = true
+		}
+
+		ib := &types.BlockInfo{
+			Height:    num,
+			Hash:      []byte(nb.Hash),
+			Finalized: finalized,
+		}
+		blocks = append(blocks, ib)
+	}
+	return blocks, nil
+
 }
 
 func (ec *EVMConsumerController) queryLatestBlocks(startKey []byte, count uint64, status finalitytypes.QueriedBlockStatus, reverse bool) ([]*types.BlockInfo, error) {
 	var blocks []*types.BlockInfo
-
+	// Can be deleted for never using
 	return blocks, nil
 }
 
 func (ec *EVMConsumerController) QueryBlock(height uint64) (*types.BlockInfo, error) {
 
-	return &types.BlockInfo{
-		Height:    height,
-		Hash:      nil,
-		Finalized: false,
-	}, nil
+	number := new(big.Int).SetUint64(height)
+
+	hexStr := Transform(number)
+
+	type Block struct {
+		Number string
+		Hash   string
+	}
+
+	var block Block
+	err := ec.evmClient.Call(&block, "eth_getBlockByNumber", hexStr, true)
+	if err != nil {
+		return nil, fmt.Errorf("can't get block by number:%s", err)
+	}
+
+	num, err := strconv.ParseUint(strings.TrimPrefix(block.Number, "0x"), 16, 64)
+	if err != nil {
+		return nil, fmt.Errorf("error:%s", err)
+	}
+
+	lastnumber, err := ec.GetLatestFinalizedNumber()
+	if err != nil {
+		return nil, fmt.Errorf("can't get latest finalized block:%s", err)
+	}
+
+	var finalized bool = false
+	if number.Cmp(lastnumber) <= 0 {
+		finalized = true
+	}
+
+	b := &types.BlockInfo{
+		Height:    num,
+		Hash:      []byte(block.Hash),
+		Finalized: finalized,
```

Comment:
I did a major refactoring in https://github.com/babylonchain/finality-provider/pull/337

basically the key point is `BlockInfo` no longer has this `Finalized` field so we don't need to fetch the finalized info every time.

i think this will have a big implication that will help simplify your code a lot.

please rebase your PR to `feat/SNA-427-QueryBlock-refactor` branch to get the latest changes.

Reply:
Replace the BlockInfo to delete Finalised